zookeeper-server/src/main/java/org/apache/zookeeper/ClientCnxn.java

```diff
@@ -971,10 +980,20 @@ void readResponse(ByteBuffer incomingBuffer) throws IOException {
          *
          * @return
          */
-        ZooKeeper.States getZkState() {
+        synchronized ZooKeeper.States getZkState() {
             return state;
         }
 
+        synchronized void setZkState(ZooKeeper.States newState) throws IOException {
+            if (!state.isAlive()) {
+                if (newState == States.CONNECTING) {
+                    throw new IOException(
```

Comment:
`if (!state.isAlive() && newState == States.CONNECTING)` ?

Reply:
Current implementation seems more structural for future evaluation because it will handle more scenarios.
But anyway I accept your suggestion. Currently your suggestion seems more compact.